### PR TITLE
Nullable reference types

### PR DESCRIPTION
**Pull request by trejjam**
Hi, I enabled Nullable reference types.

### CODE REVIEW

File: FreshdeskApi.Client/HttpClientConfigurator.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
+﻿using System;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Text;
+
+namespace FreshdeskApi.Client
+{
+    public static class HttpClientConfigurator
+    {
+        public static HttpClient ConfigureFreshdeskApi(this HttpClient httpClient, string freshdeskDomain, string apiKey)
```
Comment (trejjam) [inline FreshdeskApi.Client/HttpClientConfigurator.cs:10]: This method is useful when you want to use HttpClientFactory:
```
serviceCollection.AddHttpClient<FreshdeskClient>(client =>
{
            var freshdeskConfiguration = configurationProvider.FreshdeskConfiguration;
            client.ConfigureFreshdeskApi(freshdeskConfiguration.Domain, freshdeskConfiguration.ApiKey);
}).AddTypedClient(client => new FreshdeskClient(client));
```

**Comment by DaveTCode** [comment]
What value do you see this adding on a thin wrapper over a json api? AFAICT every property on all models ends up nullable which feels like it means we don't gain any extra clarity around the state of a given object. Can you give an example of some code which would now be able to flag up errors at compile time which wouldn't previously?

This is a genuine question and not a dig! I have a feeling that you understand this better than me (I'm yet to implement nullable reference types on any project)

**Comment by trejjam** [comment]
Hi, I see it as a library consumer. I do not read the API description if someone else did it before, and everything works. Then only from looking into a contract, I see that `Contact.Name` is filled always.

It also brings consistency: nullable Integer is `int?` and nullable String is `string?`.

Also, you can null check passed references to be non-nullable. E.g. I am preparing [PaginationConfiguration](https://github.com/trejjam/FreshdeskApiDotnet/blob/aef51fc5455c9902d6f8794f0cbd9c0ee707f80e/FreshdeskApi.Client/PaginationConfiguration.cs#L11) with optional callback. Without nullability you do not know if the callback is optional (it was the trigger to prepare this PR).

**Comment by DaveTCode** [comment]
> Hi, I see it as a library consumer. I do not read the API description if someone else did it before, and everything works. Then only from looking into a contract, I see that `Contact.Name` is filled always.
> 
> It also brings consistency: nullable Integer is `int?` and nullable String is `string?`.
> 
> Also, you can null check passed references to be non-nullable. E.g. I am preparing [PaginationConfiguration](https://github.com/trejjam/FreshdeskApiDotnet/blob/aef51fc5455c9902d6f8794f0cbd9c0ee707f80e/FreshdeskApi.Client/PaginationConfiguration.cs#L11) with optional callback. Without nullability you do not know if the callback is optional (it was the trigger to prepare this PR).

Right, so in https://github.com/DaveTCode/FreshdeskApiDotnet/pull/11/files#diff-f9d376f69a584032568e5b89de18566cR87 where you've said "this will never be null" (and then assigned it with null as a default value) I'm not so sure this is a good idea. From my experience with the Freshdesk api things that should really be invariants (e.g. Name is always set for a contact) are rarely _actually_ invariants (e.g. I've seen booleans returning strings, strings which their docs claim are mandatory being missing under certain circumstances etc! Clearly they use a database which treats schemas as optional)

But I suppose I'm not really _against_ nullable reference types so much as I don't think they're valuable on json api models where you're not in control of the data. In that case I think nearly every single entry on the api must be treated as nullable anyway.

I get your point with objects that are not part of the external API though, things like your PaginationConfiguration class do make sense.

Upshot, I think I'm fine to merge this but I'd want you to double check carefully anywhere you've said that a reference type is *not* nullable and consider whether we can really trust Freshdesk to not screw it up and miss sending that property back. My suggestion is that we can't trust them. Once you've done that I'll probably be ready to merge.

Side note: Do you want me to release a new version onto nuget.org or wait until you're finished adding features?

**Comment by trejjam** [comment]
> you've said "this will never be null" (and then assigned it with null as a default value)

I was never thinking of it this way (thanks!), more like a hint for a compiler.

> I think nearly every single entry on the api must be treated as nullable anyway.

I do not have that much experience with Freshdesk API, you are probably (and sadly) right. I am going to check these `null!`.

> Side note: Do you want me to release a new version onto nuget.org or wait until you're finished adding features?
I have the library as a git submodule for now. So I do not need it now (while I prototype our integration).

**Comment by trejjam** [comment]
Now all references used for JSON mapping are nullable

**Comment by DaveTCode** [comment]
Alright, looks good to me. I appreciate the work!

**Pull request merged**